code/game/machinery/doppler_array.dm

```diff
@@ -176,47 +173,50 @@ GLOBAL_LIST_EMPTY(doppler_arrays)
 			icon_state = "[initial(icon_state)]-off"
 			stat |= NOPOWER
 
-/obj/machinery/doppler_array/ui_interact(mob/user, ui_key = "main", var/datum/nanoui/ui = null, var/force_open = 1)
-	ui = SSnanoui.try_update_ui(user, src, ui_key, ui, force_open)
+/obj/machinery/doppler_array/tgui_interact(mob/user, ui_key = "main", datum/tgui/ui = null, force_open = FALSE, datum/tgui/master_ui = null, datum/tgui_state/state = GLOB.tgui_default_state)
+	ui = SStgui.try_update_ui(user, src, ui_key, ui, force_open)
 	if(!ui)
-		ui = new(user, src, ui_key, "doppler_array.tmpl", "Tachyon-doppler array", 500, 650)
+		ui = new(user, src, ui_key, "TachyonArray", name, 500, 600, master_ui, state)
 		ui.open()
-		ui.set_auto_update(1)
 
-/obj/machinery/doppler_array/ui_data(mob/user, ui_key = "main", datum/topic_state/state = GLOB.default_state)
-	var/data[0]
-	var/list/explosion_data = list()
-	for(var/D in logged_explosions)
-		var/datum/explosion_log/E = D
-		explosion_data += list(list(
+/obj/machinery/doppler_array/tgui_data(mob/user)
+	var/list/data = list()
+	var/list/records = list()
+	for(var/i in 1 to length(logged_explosions))
```

Comment:
Any reason for switching this to index instead of UIDs?

Reply:
The list order won't change. Only edge case is when people try to remove things at the (near) exact same time.
Using indexes here was a bit cheaper on the processing. Also less UIDs being used